accounts/tests.py

```diff
@@ -1575,3 +1576,166 @@ def test_bulk_describe_state_closed(self):
+        resp = self.client.post(reverse('flag-user', kwargs={'username': self.spammer.username}), data={
+            'object_id': object.id,
+            'flag_type': flag_type,
+        })
+        self.assertEqual(resp.status_code, 200)
+        self.assertEqual(UserFlag.objects.count(), 1)  # One flag object created
+        self.assertEqual(UserFlag.objects.first().reporting_user, reporter)  # Flag object created by reporter
+        self.assertEqual(len(mail.outbox), 0)  # No email sent
+
+        # Flag the same comment by other users so email is sent
+        for i in range(1, settings.USERFLAG_THRESHOLD_FOR_NOTIFICATION):  # Start at 1 as first flag already done
+            reporter = self.get_reporter_as_logged_in_user(i)
+            resp = self.client.post(reverse('flag-user', kwargs={'username': self.spammer.username}), data={
+                'object_id': object.id,
+                'flag_type': flag_type,
+            })
+            self.assertEqual(resp.status_code, 200)
+            self.assertEqual(UserFlag.objects.count(), i + 1)  # Now we have more flags
+            self.assertEqual(UserFlag.objects.all().order_by('id')[i].reporting_user,
+                             reporter)  # Flag object created by reporter
+
+            if i == settings.USERFLAG_THRESHOLD_FOR_NOTIFICATION - 1:  # Last iteration
+                self.assertEqual(len(mail.outbox), 1)  # Notification email sent
+                self.assertTrue("[freesound] Spam/offensive report for user" in mail.outbox[0].subject)
+                self.assertTrue("has been reported" in mail.outbox[0].body)
+
+        # Continue flagging object until it reaches blocked state
+        for i in range(settings.USERFLAG_THRESHOLD_FOR_NOTIFICATION,
+                       settings.USERFLAG_THRESHOLD_FOR_AUTOMATIC_BLOCKING):  # Start at 1 as first flag already done
+            reporter = self.get_reporter_as_logged_in_user(i)
+            resp = self.client.post(reverse('flag-user', kwargs={'username': self.spammer.username}), data={
+                'object_id': object.id,
+                'flag_type': flag_type,
+            })
+            self.assertEqual(resp.status_code, 200)
+            self.assertEqual(UserFlag.objects.count(), i + 1)  # Now we have more flags
+            self.assertEqual(UserFlag.objects.all().order_by('id')[i].reporting_user,
+                             reporter)  # Flag object created by reporter
+
+            if i == settings.USERFLAG_THRESHOLD_FOR_AUTOMATIC_BLOCKING - 1:  # Last iteration
+                self.assertEqual(len(mail.outbox), 2)  # New notification email sent
+                self.assertTrue("[freesound] Spam/offensive report for user" in mail.outbox[1].subject)
+                self.assertTrue("has been blocked" in mail.outbox[1].body)
+
+            else:
+                self.assertEqual(len(mail.outbox), 1)  # No new wmail sent
+
+        # Flag the object again and no new notification emails are sent
```

Comment:
I would have this as a separate test, and only do it for one object type (if the logic is the same for all of them)

Reply:
Hmm, what I'm testing here is that a first email is sent when a user has object(s) flagged by at least 3 users, and that a second email is sent when flags come from 6 distinct users (and the reported user is blocked). Then no more emails should be sent. That last bit of the test is to check that no more emails are sent so to me it is part of the same thing right? I keep on flagging user's content and checking if an email was sent or not.